### DISCUSSION TITLE
rspack如何自定义.env.[mode]呢？

### DISCUSSION THREAD

**Discussion by limuen** [web-infra-dev/rspack#8811] [Q&A] [answered]
"build:qa": "rspack build --env-mode qa"

我的.env.qa文件 
.env.qa

npm run build:qa

> rspack build --env-mode qa01

无法识别这些选项：env-mode, envMode

**Comment by inottn** [comment]
You can refer to the implementation of Rsbuild.

**Comment by chenjiahan** [comment]
Use this plugin: https://github.com/mrsteele/dotenv-webpack

**Reply by limuen** [reply]
还是又回到这里请教了，我现在在plugins里面配置了
new DotenvWebpack({
    path: `.env.${process.env.NODE_ENV}`,
    safe: false,
    allowEmptyValues: true,
    systemvars: true,
    silent: false,
    defaults: false,
  }),
  
  并且注释掉了rspack.DefinePlugin
  // new rspack.DefinePlugin({
  //   "process.env": JSON.stringify(filterEnv)
  // }),
  
  环境也成功的都拿到了，但是终端会给个警告
  WARNING in ⚠ DefinePlugin:
  │ Conflicting values for 'process.env.NODE_ENV' ("\"qa-01\"" !== "\"development\"")
  
  我不想在ignoreWarnings里面给添加上process.env的选项，请问有其他方法可以解决这个警告吗
  
  @chenjiahan

**Reply by chenjiahan** [reply, marked as answer]
Try to disable the default `optimization.nodeEnv`:

https://rspack.dev/config/optimization#optimizationnodeenv

**Reply by limuen** [reply]
感谢 🙏

**Comment by sweetwisdom** [comment]
[参考](https://v0.rsbuild.dev/zh/guide/advanced/env-vars#%E6%96%87%E4%BB%B6%E7%B1%BB%E5%9E%8B)
 
 **Vue CLI 那样的多 mode / 多环境支持**（如 `development`、`production`、`staging`、`test` 等），即：

- 通过 `--mode xxx` 指定不同环境
- 自动加载 `.env`、`.env.[mode]`、`.env.[mode].local` 等文件
- 把环境变量注入到代码中（类似 `VUE_APP_xxx` 或 `import.meta.env`）

### 1. 推荐使用 Rsbuild（基于 Rspack）

Rspack 本身是底层引擎，**Rsbuild 是官方推荐的上层工具**，它已经内置了类似 Vue CLI 的多环境支持，使用起来最简单。

```bash
# 安装
npm install @rsbuild/core @rsbuild/plugin-vue   # 或其他框架插件
```

**rsbuild.config.ts**（或 .js）：

```ts
import { defineConfig, loadEnv } from '@rsbuild/core';

export default defineConfig({
  // 基础配置
  source: {
    entry: { index: './src/main.ts' },
  },
  // ... 其他配置
});
```

**使用多 mode**：

```bash
# 开发
npx rsbuild dev --env-mode development

# 构建 production
npx rsbuild build --env-mode production

# 自定义 staging 环境
npx rsbuild build --env-mode staging
```

Rsbuild 会自动按优先级加载环境变量文件：

- `.env`
- `.env.local`
- `.env.[mode]`
- `.env.[mode].local`

在代码中可以直接使用：

```ts
console.log(import.meta.env.MODE);           // 'staging'
console.log(import.meta.env.DEV);            // true/false
console.log(import.meta.env.PROD);

// 自定义变量（推荐前缀 BASE_ 或 PUBLIC_ 等）
console.log(import.meta.env.BASE_API_URL);
```

**自定义注入变量**（类似 Vue 的 `VUE_APP_`）：

```ts
import { defineConfig, loadEnv } from '@rsbuild/core';

export default defineConfig(({ mode }) => {
  const { publicVars } = loadEnv({ mode, prefixes: ['VUE_APP_', 'BASE_'] });

  return {
    source: {
      define: publicVars,   // 自动注入
    },
  };
});
```
